sdk/src/logs/batch_log_processor.cc

```diff
@@ -0,0 +1,225 @@
+{
+  if (is_shutdown_.load() == true)
+  {
+    return false;
+  }
+
+  is_force_flush_ = true;
+
+  // Keep attempting to wake up the worker thread
+  while (is_force_flush_.load() == true)
+  {
+    cv_.notify_one();
+  }
+
+  // Now wait for the worker thread to signal back from the Export method
+  std::unique_lock<std::mutex> lk(force_flush_cv_m_);
+  while (is_force_flush_notified_.load() == false)
+  {
+    force_flush_cv_.wait(lk);
+  }
+
+  // Notify the worker thread
+  is_force_flush_notified_ = false;
+  return true;
+}
+
+void BatchLogProcessor::DoBackgroundWork()
+{
+  auto timeout = schedule_delay_millis_;
+
+  while (true)
+  {
+    // Wait for `timeout` milliseconds
+    std::unique_lock<std::mutex> lk(cv_m_);
+    cv_.wait_for(lk, timeout);
+
+    if (is_shutdown_.load() == true)
+    {
+      DrainQueue();
+      return;
+    }
+
+    bool was_force_flush_called = is_force_flush_.load();
+
+    // Check if this export was the result of a force flush.
+    if (was_force_flush_called == true)
+    {
+      // Since this export was the result of a force flush, signal the
+      // main thread that the worker thread has been notified
+      is_force_flush_ = false;
+    }
+    else
+    {
+      // If the buffer was empty during the entire `timeout` time interval,
+      // go back to waiting. If this was a spurious wake-up, we export only if
+      // `buffer_` is not empty. This is acceptable because batching is a best
+      // mechanism effort here.
+      if (buffer_.empty() == true)
+      {
+        continue;
+      }
+    }
+
+    auto start = std::chrono::steady_clock::now();
+
+    Export(was_force_flush_called);
+
+    auto end      = std::chrono::steady_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
+
+    // Subtract the duration of this export call from the next `timeout`.
+    timeout = schedule_delay_millis_ - duration;
+  }
+}
+
+void BatchLogProcessor::Export(const bool was_force_flush_called)
+{
+  std::vector<std::unique_ptr<LogRecord>> records_arr;
+
+  size_t num_records_to_export;
+
+  if (was_force_flush_called == true)
+  {
+    num_records_to_export = buffer_.size();
+  }
+  else
+  {
+    num_records_to_export =
+        buffer_.size() >= max_export_batch_size_ ? max_export_batch_size_ : buffer_.size();
+  }
+
+  buffer_.Consume(
+      num_records_to_export, [&](CircularBufferRange<AtomicUniquePtr<LogRecord>> range) noexcept {
+        range.ForEach([&](AtomicUniquePtr<LogRecord> &ptr) {
+          std::unique_ptr<LogRecord> swap_ptr = std::unique_ptr<LogRecord>(nullptr);
+          ptr.Swap(swap_ptr);
+          records_arr.push_back(std::unique_ptr<LogRecord>(swap_ptr.release()));
+          return true;
+        });
+      });
+
+  ExportResult export_status = exporter_->Export(records_arr);
+  if (export_status != ExportResult::kSuccess)
+  {
+    std::cerr << "[Batch Log Processor] Export failed \n";
```

Comment:
We don't have any error handling mechanism yet. Better to remove this error message, and update the existing github ticket #408 for handling this error,

Reply:
Sure - will make that change